app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt

```diff
@@ -78,43 +65,44 @@ class SearchFragment : Fragment(), BackHandler {
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
-        sessionId = SearchFragmentArgs.fromBundle(arguments!!).sessionId
-        isPrivate = (activity as HomeActivity).browsingModeManager.isPrivate
+        val session = arguments
+            ?.let(SearchFragmentArgs.Companion::fromBundle)
+            ?.let { it.sessionId }
+            ?.let(requireComponents.core.sessionManager::findSessionById)
 
-        val session = sessionId?.let { requireComponents.core.sessionManager.findSessionById(it) }
         val view = inflater.inflate(R.layout.fragment_search, container, false)
         val url = session?.url ?: ""
 
-        toolbarComponent = ToolbarComponent(
-            view.toolbar_component_wrapper,
-            ActionBusFactory.get(this),
-            sessionId,
-            isPrivate,
-            true,
-            view.search_engine_icon,
-            FenixViewModelProvider.create(
-                this,
-                ToolbarViewModel::class.java
-            ) {
-                ToolbarViewModel(SearchState(url, session?.searchTerms ?: "", isEditing = true))
-            }
-        ).also {
-            // Remove background from toolbar view since it conflicts with the search UI.
-            it.uiView.view.background = null
-            it.uiView.view.layoutParams.height = CoordinatorLayout.LayoutParams.MATCH_PARENT
-        }
+        val viewModel = StateViewModel.get(
+            this,
+            SearchState(
+                query = url,
+                showShortcutEnginePicker = false,
+                searchEngineSource = SearchEngineSource.Default(
+                    requireComponents.search.searchEngineManager.getDefaultSearchEngine(requireContext())
+                ),
+                showSuggestions = Settings.getInstance(requireContext()).showSearchSuggestions,
+                showVisitedSitesBookmarks = Settings.getInstance(requireContext()).shouldShowVisitedSitesBookmarks,
+                session = session
+            )
+        )
 
-        awesomeBarComponent = AwesomeBarComponent(
-            view.search_layout,
-            ActionBusFactory.get(this),
-            FenixViewModelProvider.create(
-                this,
-                AwesomeBarViewModel::class.java
-            ) {
-                AwesomeBarViewModel(AwesomeBarState("", false))
-            }
+        searchStore = Store(
+            viewModel.state,
+            ::searchStateReducer
         )
-        ActionBusFactory.get(this).logMergedObservables()
+
+        searchStore.observe(this) { viewModel.update(it) }
+
+        searchInteractor = SearchInteractor(
+            activity as HomeActivity,
+            findNavController(),
+            searchStore
+        )
+
+        toolbarView = ToolbarView(view.toolbar_component_wrapper, searchInteractor, ::historyStorageProvider)
```

Comment:
I am wondering about the `::historyStorageProvider`. It doesn't seem to be used anywhere else in the fragment. Could we just pass the actual `HistoryStorage` to the `ToolbarView` and decide there whether or not to use it based on `Settings.getInstance(requireContext()).shouldShowVisitedSitesBookmarks`. Don't want to pile up too many refactorings though so this could be a follow-up if we decide it's worth it.